# Multiomics analysis of aggregative multicellularity
Source code used to analyse the multi-omics datasets generated during *Dictyostelium discoideum* multicellular development, as well as any figures generated. The code can be found in **Omics_figures_v2.Rmd**, and generated figures and tables in the plots and tables folders, respectively. Input data and annotations are located in data and dictybase_20200923 folders, respectively. A compiled output file of the full analysis with all plots is available in pdf format.

## Data
The proteomics data is available at MassIVE, with identifier [MSV000093620](https://doi.org/doi:10.25345/C5H12VJ75)  
The transcriptomics data is available at GEO, with identifier [GSE249880](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE249880)

The transcriptomics and proteomics data can also be accessed interactively at https://edelbroek2024.serve.scilifelab.se/app/edelbroek2024

Protein quantifications from Kelly et al. proteomics data[^1], counts from Rosengarten et al. transcriptomics data[^2], used milestone genes[^3] and developmental genes[^4] are in the data folder.

[^1]: Kelly, B., Carrizo, G.E., Edwards-Hicks, J., Sanin, D.E., Stanczak, M.A., Priesnitz, C., Flachsmann, L.J., Curtis, J.D., Mittler, G., Musa, Y., et al. (2021). Sulfur sequestration promotes multicellularity during nutrient limitation. Nature 591, 471–476. https://doi.org/10.1038/s41586-021-03270-3
[^2]: Rosengarten, R.D., Santhanam, B., Fuller, D., Katoh-Kurasawa, M., Loomis, W.F., Zupan, B., and Shaulsky, G. (2015). Leaps and lulls in the developmental transcriptome of Dictyostelium discoideum. BMC Genomics 16, 294. https://doi.org/10.1186/s12864-015-1491-7
[^3]: Katoh-Kurasawa, M., Hrovatin, K., Hirose, S., Webb, A., Ho, H.-I., Zupan, B., and Shaulsky, G. (2021). Transcriptional milestones in Dictyostelium development. Genome Res 31, 1498–1511. https://doi.org/10.1101/gr.275496.121
[^4]: Loomis, W.F. (2015). Genetic control of morphogenesis in Dictyostelium. Developmental Biology 402, 146–161. https://doi.org/10.1016/j.ydbio.2015.03.016
